File: apps/backend/src/portfolio/Portfolio.ts
import {
  CreateDateColumn,
  Entity,
  JoinColumn,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { PortfolioAsset } from './PortfolioAsset';
import { UUID } from '@mid/shared';
import { Strategy } from '../strategy/Strategy';
import { User } from '../user/User';

@Entity()
export class Portfolio {
  @PrimaryGeneratedColumn('uuid')
  id: UUID;

  @ManyToOne(() => Strategy, { lazy: true, nullable: false })
  @JoinColumn({ name: 'strategy_id' })
  strategy: Strategy;

  @OneToMany(
    () => PortfolioAsset,
    (portfolioAsset) => portfolioAsset.portfolio,
    {
      cascade: true,
      eager: true,
      orphanedRowAction: 'delete',
    },
  )
  portfolioAssets: PortfolioAsset[];

  @CreateDateColumn({ type: 'timestamp', default: () => 'CURRENT_TIMESTAMP' })
  createdAt: Date;

  @ManyToOne(() => User, {
    eager: true,
    nullable: false,
  })
  user: User;
}

File: backend/src/main/kotlin/com/kyoulho/mid/portfolio/Portfolio.kt
package com.kyoulho.mid.portfolio

import com.kyoulho.mid.strategy.Strategy
import com.kyoulho.mid.user.entity.MidUser
import jakarta.persistence.*
import org.hibernate.annotations.CreationTimestamp
import java.time.LocalDateTime

@Entity
data class Portfolio(
        @Id
        @GeneratedValue(strategy = GenerationType.UUID)
        val id: String? = null,

        @ManyToOne(fetch = FetchType.LAZY, optional = false)
        @JoinColumn(name = "strategy_id", nullable = false)
        val strategy: Strategy,

        @OneToMany(mappedBy = "portfolio", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.EAGER)
        val portfolioAssets: List<PortfolioAsset> = mutableListOf(),

        @CreationTimestamp
        @Column(name = "created_at", nullable = false, updatable = false)
        val createdAt: LocalDateTime? = null,

        @ManyToOne(fetch = FetchType.EAGER, optional = false)
        @JoinColumn(name = "user_id", nullable = false)
        val user: MidUser
)
